Universal/Sniffs/CodeAnalysis/MixedBooleanOperatorSniff.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/**
+ * PHPCSExtra, a collection of sniffs and standards for use with PHP_CodeSniffer.
+ *
+ * @package   PHPCSExtra
+ * @copyright 2021 WoltLab GmbH, 2023 PHPCSExtra Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCSStandards/PHPCSExtra
+ */
+
+namespace PHPCSExtra\Universal\Sniffs\CodeAnalysis;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHPCSUtils\BackCompat\BCFile;
+
+/**
+ * Forbid mixing different binary boolean operators within a single expression without making precedence
+ * clear using parentheses.
+ *
+ * @link https://github.com/squizlabs/PHP_CodeSniffer/pull/3205
+ * @link https://github.com/php-fig/per-coding-style/issues/22
+ *
+ * @since 1.2.0
+ */
+final class MixedBooleanOperatorSniff implements Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 1.2.0
+     *
+     * @return array<int|string>
+     */
+    public function register()
+    {
+        return [
+            \T_BOOLEAN_OR,
+            \T_BOOLEAN_AND,
+            \T_LOGICAL_AND,
+            \T_LOGICAL_OR,
+            \T_LOGICAL_XOR,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 1.2.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token
+     *                                               in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+        $token  = $tokens[$stackPtr];
+
+        $start = BCFile::findStartOfStatement($phpcsFile, $stackPtr);
+
+        $valid = $token['code'];
+
+        $previous = $phpcsFile->findPrevious(
+            \array_filter(
+                \array_merge(
+                    $this->register(),
+                    [\T_INLINE_THEN, \T_INLINE_ELSE]
+                ),
+                function ($token) use ($valid) {
+                    return $token !== $valid;
+                }
+            ),
```

Comment:
Sniffs should be very conscious about performance as they are called every single time a certain token is encountered.
This also means that this bit of code will be run _every single time_ the tokens are encountered, which seems unnecessary as we're basically talking a largely fixed array.

A typical pattern to improve performance for this code would be as follows (pseudo code to explain):
```php
    private $searchTargets;

    public function register()
    {
        // Set $searchTargets only once.
        $this->searchTargets = Tokens::$booleanOperators;
        $this->searchTargets[\T_INLINE_THEN] = \T_INLINE_THEN;
        $this->searchTargets[\T_INLINE_ELSE] = \T_INLINE_ELSE;

        return Tokens::$booleanOperators;
    }

    public function process()
    {
        ...
        $valid = $token['code'];
        $search = $this->searchTargets;
        unset($search[$valid]);

        $previous = $phpcsFile->findPrevious($search ....);
        ...
    }
```

Reply:
Thanks, I could simplify this and improve the output at the same time:

```
if (true || true && true && true && true); // Not OK.
```

previously this reported each `&&` separately, now only the first one is reported. This should result in less now, because the developer will need to look at the expression anyway.

Reply:
Generally speaking, hiding one error "behind" another is a no-no for sniffs (as people will fix something and then, surprise!, a new error along the same lines pops up).

Having said that, in this case I think it makes sense to only flag the first operator which is different, so all good.